javalib/src/main/scala/java/util/Formatter.scala

```diff
@@ -79,1587 +91,989 @@ class Formatter(
-          return result.toString()
-        }
-      } else {
-        if (formatToken.isFlagSet(FormatToken.FLAG_SPACE)) {
-          result.insert(0, ' ')
-          startIndex += 1
-        }
-        if (formatToken.isFlagSet(FormatToken.FLAG_ADD)) {
-          result.insert(0, floatUtil.getAddSign())
-          startIndex += 1
+        arg match {
+          case arg: Byte  => padAndSendWithHexInt(arg & 0xFF)
+          case arg: Char  => padAndSendWithHexInt(arg & 0xFFFF)
+          case arg: Short => padAndSendWithHexInt(arg & 0xFFFF)
+          case arg: Int   => padAndSendWithHexInt(arg)
+          case arg: Long =>
+            padAndSendToDest(
+              RootLocaleInfo,
+              flags,
+              width,
+              prefix,
+              applyNumberUpperCase(flags, java.lang.Long.toHexString(arg)))
+          case arg: BigInteger =>
+            formatNumericString(RootLocaleInfo,
+                                flags,
+                                width,
+                                arg.toString(16),
+                                prefix)
+          case _ =>
+            formatNullOrThrowIllegalFormatConversion()
         }
-      }
+        validateFlags(flags,
+                      conversion,
+                      invalidFlags = InvalidFlagsForOctalAndHex)
+      case 'e' | 'E' =>
+        validateFlags(flags, conversion, invalidFlags = UseGroupingSeps)
+        efgCommon(computerizedScientificNotation _)
+
+      case 'g' | 'G' =>
+        validateFlags(flags, conversion, invalidFlags = AltFormat)
+        efgCommon(generalScientificNotation _)
+
+      case 'f' =>
+        validateFlags(flags, conversion, invalidFlags = 0)
+        efgCommon(decimalNotation _)
+
+      case '%' =>
+        validateFlagsForPercentAndNewline(flags,
+                                          conversion,
+                                          invalidFlags =
+                                            AllWrittenFlags & ~LeftAlign)
+        rejectPrecision()
+        if (flags.leftAlign && width < 0)
+          throw new MissingFormatWidthException("%-%")
+        padAndSendToDestNoZeroPad(flags, width, "%")
+
+      case 'n' =>
+        rejectPrecision()
+        if (width >= 0)
+          throw new IllegalFormatWidthException(width)
+        validateFlagsForPercentAndNewline(flags,
+                                          conversion,
+                                          invalidFlags = AllWrittenFlags)
+
+        sendToDest("\n")
+
+      // todo case 't' | 'T' => date/time
+      // todo case 'a' | 'A' => floating point formatted as hex
+      case _ =>
+        throw new UnknownFormatConversionException(conversion.toString)
+    }
+  }
 
-      val firstChar = result.charAt(0)
-      if (formatToken.isFlagSet(FormatToken.FLAG_ZERO) && (firstChar == floatUtil
-            .getAddSign() || firstChar == floatUtil.getMinusSign()))
-        startIndex = 1
+  @inline private def validateFlags(flags: Flags,
```

Comment:
Put annotations on a separate line if the whole method header does not fit on one line:
```suggestion
  @inline
  private def validateFlags(flags: Flags,
```